Test for pbs files on slurm

aalitaiga opened the pull request.

aalitaiga committed:
Added test for slurm integration

aalitaiga committed:
New test for priority

aalitaiga committed:
Added gres + memory tests

aalitaiga committed:
Refactored tests

aalitaiga committed:
small update

bouthilx commented on smartdispatch/tests/pbs_slurm_test.py:
```diff
@@ -0,0 +1,133 @@
+
+#PBS -N arrayJob
+#PBS -o arrayJob_%A_%a.out
+#PBS -t 1-2
```
We can ignore arrays as long as SmartDispatch is not using them.

bouthilx commented on smartdispatch/tests/pbs_slurm_test.py:
```diff
@@ -0,0 +1,133 @@
+        )
+
+#
+# pbs_string2 = """\
```
Why a second PBS template?

bouthilx commented on smartdispatch/tests/pbs_slurm_test.py:
```diff
@@ -0,0 +1,133 @@
+            text_file.write(command)
+        process = Popen("sbatch test.pbs", stdout=PIPE, stderr=PIPE, shell=True)
+        stdout, stderr = process.communicate()
+        assert_true("Submitted batch job" in stdout)
```
You can define the function as a Method of TestSlurm and use ```self.assertIn``` rather than ```assert_true```

bouthilx commented on smartdispatch/tests/pbs_slurm_test.py:
```diff
@@ -0,0 +1,133 @@
+        stdout, stderr = process.communicate()
+        job_params = [c.strip() for c in stdout.split("\n")[1:] if c != '']
+        # import ipdb; ipdb.set_trace()
+        assert_true(all(p == param for p in job_params))
```
Again, this would give a more detailed error message ```self.assertSequenceEqual(job_params, (param for _ in job_params))```

bouthilx commented on smartdispatch/tests/pbs_slurm_test.py:
```diff
@@ -0,0 +1,133 @@
+
+    def tearDown(self):
+        process = Popen("rm *.out test.pbs", stdout=PIPE, stderr=PIPE, shell=True)
+        stdout, stderr = process.communicate()
```
I think it'd be better to use glob instead of Popen. 

```python
for file_name in (glob.glob('*.out') + ["test.pbs"]):
    os.remove(file_name)
```

aalitaiga committed:
Python3 compatibility + PR comments

bouthilx commented on smartdispatch/tests/pbs_slurm_test.py:
```diff
@@ -0,0 +1,114 @@
+#SBATCH --job-name=arrayJob
+#SBATCH --output=arrayJob_%A_%a.out
+#SBATCH --time=01:00:00
+#SBATCH --gres=gpu
```
-l naccelerators=x will give gpu:x with ```squeue -o gres```, so you should use --gres=gpu:x in sbatch file too. You should not test for 1 only, give it a random integer too.

bouthilx commented on smartdispatch/tests/pbs_slurm_test.py:
```diff
@@ -0,0 +1,114 @@
+#SBATCH --output=arrayJob_%A_%a.out
+#SBATCH --time=01:00:00
+#SBATCH --gres=gpu
+#SBATCH --constraint=gpu6gb
```
gpu:x and constraint should be formatable in sbatch string since we want to try different values.

aalitaiga committed:
Fixed naccelerators issue

aalitaiga committed:
Updated tests to skip on Graham and Cedar smartdispatch modified to handle slurm clusters

bouthilx commented on smartdispatch/tests/pbs_slurm_test.py:
```diff
@@ -48,26 +46,25 @@ def tearDown(self):
+        output_array = output_array or param_array
+        for param, output in zip(param_array, output_array):
+            com = pbs_string.format(
+                string.format(command.format(param))
```
Just for clarity:
command -> command_template
com -> param_command

bouthilx commented on smartdispatch/job_generator.py:
```diff
@@ -171,3 +181,13 @@ def _add_cluster_specific_rules(self):
+
+    def _add_cluster_specific_rules(self):
+        for pbs in self.pbs_list:
+            node_resource = pbs.resources.pop('nodes')
```
It doesn't seem wise to pop ['nodes'] if there is other information than gpus and ppn. I would rather substitute "gpus=[0-9]+:" with "". However we would need to check if resources['nodes'] becomes empty afterwards and remove it if so. It is very likely to happen, but it is necessary if we want to keep other informations when present.

bouthilx commented on smartdispatch/job_generator.py:
```diff
@@ -171,3 +181,13 @@ def _add_cluster_specific_rules(self):
+
+    def _add_cluster_specific_rules(self):
+        for pbs in self.pbs_list:
+            node_resource = pbs.resources.pop('nodes')
```
It doesn't seem wise to pop ['nodes'] if there is other information than gpus and ppn. I would rather substitute "gpus=[0-9]+:" with "". However we would need to check if resources['nodes'] becomes empty afterwards and remove it if so. It is very likely to happen, but it is necessary if we want to keep other informations when present.

aalitaiga replied:
If go from "nodes=1:ppn=2:gpus=2" to "nodes=1" is it okay or I should remove nodes completely in this case?

bouthilx replied:
It is ok

bouthilx commented on smartdispatch/pbs.py:
```diff
@@ -62,6 +64,18 @@ def add_options(self, **options):
+        Parameters
+        ----------
+        **options : dict
+            each key is the name of a SBATCH option (see `Options`)
```
There is no Options section in this docstring

bouthilx commented on smartdispatch/pbs.py:
```diff
@@ -62,6 +64,18 @@ def add_options(self, **options):
+        """
+
+        for option_name, option_value in options.items():
+            self.sbatch_options[option_name] = option_value
```
We would need to add back the "-"/"--" here if we drop it in ```add_sbatch_flags```. 
```python
if len(option_name) == 1:
    "-" + option_name 
else:
    "--" + option_name
```

bouthilx replied:
Isn't there a duplicate at line 81? You can change it to option_name = "-" + option_name and leave line 81 like it is.

bouthilx commented on smartdispatch/job_generator.py:
```diff
@@ -73,6 +73,16 @@ def add_pbs_flags(self, flags):
+
+        for flag in flags:
+            split = flag.find('=')
+            options[flag[:split]] = flag[split+1:]
```
It would be more consistent with the PBS options treatment if we drop the "-"/"--" in front of the option names. I would also raise a ValueError if neither are present.

bouthilx commented on smartdispatch/tests/pbs_slurm_test.py:
```diff
@@ -39,6 +39,13 @@
 """
 
+# Checking which cluster is running the tests first
+process = Popen("sacctmgr list cluster", stdout=PIPE, stderr=PIPE, shell=True)
```
would subprocess.check_output be simpler to use? It's used widely in this repo.

bouthilx replied:
Also I would write functions for qstat and slurm cluster detections. Something like 
```python
def get_qstat_cluster_name():
    """
    ... Some doc ...

    Raises
    ------
    OSError
        The cluster may not have qstat installed.
    """
    output = Popen(["qstat", "-B"], stdout=PIPE).communicate()[0]
    return output.split('\n')[2].split(' ')[0]


def get_slurm_cluster_name():
    """
    ... Some doc ...
    Raises
    ------
    OSError
        The cluster may not have slurm installed.
    """
    # ... Some code ...
```

Those would be used in ```detect_cluster()```

bouthilx commented on smartdispatch/utils.py:
```diff
@@ -115,6 +115,7 @@ def detect_cluster():
         output = Popen(["qstat", "-B"], stdout=PIPE).communicate()[0]
     except OSError:
         # If qstat is not available we assume that the cluster is unknown.
+        # TODO: handle MILA + CEDAR + GRAHAM
```
You can insert the function ```get_slurm_cluster_name()``` here for reference, but let it still return None. We don't want to add support for specific clusters now. That will be another ticket.

bouthilx commented:
@aalitaiga You should rename pbs_slurm_test.py otherwise Travis will keep reporting failures. We could add --ignore-files=<name> in .travis.yml but we don't want those tests to run locally neither. They are only usefull if we want to tests new settings on the slurm configuration, so they should only be ran alone. A name like verify_slurms_pbs_wrapper.py would be self-explanatory.

aalitaiga committed:
Cleaned code with PR feedback

bouthilx commented on smartdispatch/job_generator.py:
```diff
@@ -78,7 +78,12 @@ def add_sbatch_flags(self, flags):
             split = flag.find('=')
-            options[flag[:split]] = flag[split+1:]
+            if flag.startswith('--'):
+                options[flag[2:split]] = flag[split+1:]
```
This would be simpler.

```python
if not flag.startswith("-"):
  raise ValueError("Invalid SBATCH flag ({})".format(flag))

options[flag[:split].lstrip("-")] = flag[split+1:]
```

bouthilx commented on smartdispatch/job_generator.py:
```diff
@@ -182,12 +187,13 @@ def _add_cluster_specific_rules(self):
+            gpus = re.match(".*gpus=([0-9]+)", pbs.resources['nodes']).group(1)
+            ppn = re.match(".*ppn=([0-9]+)", pbs.resources['nodes']).group(1)
+            pbs.resources['nodes'] = re.sub("ppn=[0-9]+", "", pbs.resources['nodes'])
+            pbs.resources['nodes'] = re.sub(":gpus=[0-9]+", "", pbs.resources['nodes'])
```
You don't look for empty pbs.resources['nodes'] because there is always nodes=1 remaining? That would be fine in my opinion, I just want to be sure to understand for which reason you do not check.

aalitaiga replied:
Yes it is the reason

bouthilx commented on smartdispatch/utils.py:
```diff
@@ -115,7 +115,7 @@ def detect_cluster():
     except OSError:
         # If qstat is not available we assume that the cluster is unknown.
-        # TODO: handle MILA + CEDAR + GRAHAM
+        cluster_name = get_slurm_cluster_name()
```
Please leave the TODO there. Otherwise, it is not obvious why it still returns None.

bouthilx commented on smartdispatch/utils.py:
```diff
@@ -131,6 +131,11 @@ def detect_cluster():
         cluster_name = "hades"
     return cluster_name
 
+def get_slurm_cluster_name():
```
We need a unit test for this

aalitaiga replied:
There's no unittest for 'detect_cluster' should I make one for 'get_cluster_name'?

bouthilx replied:
There should be one for each functions.

bouthilx commented on smartdispatch/job_generator.py:
```diff
@@ -73,6 +73,21 @@ def add_pbs_flags(self, flags):
             pbs.add_resources(**resources)
             pbs.add_options(**options)
 
+    def add_sbatch_flags(self, flags):
```
We need a unit test for add_sbatch_flags

bouthilx commented on smartdispatch/job_generator.py:
```diff
@@ -171,3 +186,14 @@ def _add_cluster_specific_rules(self):
+
+class SlurmJobGenerator(JobGenerator):
+
+    def _add_cluster_specific_rules(self):
```
We need a unit test for this

bouthilx commented on scripts/smart-dispatch:
```diff
@@ -206,6 +209,7 @@ def parse_arguments():
 
     parser.add_argument('-p', '--pool', type=int, help="Number of workers that will be consuming commands. Default: Nb commands")
     parser.add_argument('--pbsFlags', type=str, help='ADVANCED USAGE: Allow to pass a space seperated list of PBS flags. Ex:--pbsFlags="-lfeature=k80 -t0-4"')
+    parser.add_argument('--sbatchFlags', type=str, help='ADVANCED USAGE: Allow to pass a space seperated list of SBATCH flags. Ex:--sbatchFlags="-qos=high --output=file.out"')
```
We need a unit test for this.

bouthilx commented:
We need unit-tests for 
- [x] add_sbatch_flags
- [x] add_sbatch_options
- [x] ~~command line option~~
- [x] job generator pbs file replacements (pnn, gpus, etc)

Make sure to test that what should pass, pass and what should fail fails. We need to try many different things to test the command line options. For instance, we need to try adding bad options or adding options already included by job generator. We need to make sure to test every cracks as this seems to me as an error-prone setup (replacing options for a framework by options a wrapper of another framework will recognize as what we want it to be for the latter framework :sweat:).

aalitaiga committed:
Updated tests

bouthilx commented on smartdispatch/job_generator.py:
```diff
@@ -79,11 +79,13 @@ def add_sbatch_flags(self, flags):
-            elif flag.startswith('-'):
-                options[flag[1:split]] = flag[split+1:]
+                if split == -1:
+                    raise ValueError("Invalid SBATCH flag ({})".format(flag))
```
We know it's invalid probably because it should be a single - or the = is missing. The error message should give a hint about it.

bouthilx commented on smartdispatch/job_generator.py:
```diff
@@ -79,11 +79,13 @@ def add_sbatch_flags(self, flags):
+                    raise ValueError("Invalid SBATCH flag ({})".format(flag))
+                options[flag[:split].lstrip("-")] = flag[split+1:]
+            elif flag.startswith('-') and split == -1:
+                options[flag[1:2]] = flag[2:]
```
simply flag[1]? :stuck_out_tongue:

bouthilx commented on smartdispatch/tests/test_job_generator.py:
```diff
@@ -129,6 +130,32 @@ def test_add_pbs_flags_invalid(self):
+    def test_add_sbatch_flag_invalid(self):
+        invalid_flags = ["--qos high", "gpu", "-lfeature=k80"]
+        for flag in invalid_flags:
+            assert_raises(ValueError, self._test_add_sbatch_flags, "--qos high")
```
you pass the same flag each time.

bouthilx commented on smartdispatch/tests/test_job_generator.py:
```diff
@@ -244,6 +271,28 @@ def test_pbs_split_2_job_nb_commands(self):
+        self.pbs = job_generator.pbs_list
+
+    def test_ppn_ncpus(self):
+        assert_true("ppn" not in str(self.pbs[0]))
```
Don't forget to make a dummy mock of self._add_cluster_specific_rules() to test that otherwise ppn was indeed present in the PBS and ncpus was not.

aalitaiga committed:
Updated tests using mock

bouthilx committed:
Refactor detect_cluster tests

aalitaiga committed:
Small changes in TestSlurmQueue

bouthilx committed:
Fix add_sbatch_option bug

Why:

For each option, add_sbatch_option would add the option in both the form
--[OPTION_NAME] and [OPTION_NAME].

bouthilx committed:
Refactor SlurmJobGenerator

It will need many conversions, not only on resources, so better make it
clean.

bouthilx committed:
Remove queue name for Slurm clusters

Slurm has no queues, so PBS option -q is invalid and non-convertible.

bouthilx committed:
Replace PBS_JOBID with SLURM_JOB_ID

$PBS_JOBID was used to set the stdout/err of the job as well as in the
commands. Replace them with $SLURM_JOB_ID.

Also, workers were accessing os.environ[PBS_JOBID] so we added a second
fetch on SLURM_JOB_ID in case os.environ[PBS_JOBID] gave undefined.

bouthilx committed:
Add PBS_FILENAME definition to pbs.prolog

Slurm cannot be passed environment variables defined locally on
command-line like PBS_FILENAME is. To bypass this, we add a definition
in the prolog, making PBS_FILENAME available to all commands and epilog.

NOTE: We leave PBS_FILENAME definition in command-line too such that any
user using $PBS_FILENAME inside a custom pbsFlag can still do so.

bouthilx committed:
Fix env var export option for Slurm

PBS options -V is not converted properly to SBATCH --export ALL.
We remove it and replace it with --export=ALL is the sbatch options.

bouthilx committed:
Adapt PBS_WALLTIME for slurm

Slurm does not have a equivalent environment variable set like
PBS_WALLTIME. To avoid confusion, all variables PBS_WALLTIME are renamed
to SBATCH_TIMELIMIT (the environment variable one would use to set --time
with sbatch). As SBATCH_TIMELIMIT is not set automatically, we add it to
the prolog to make it available to all commands and epilog.

NOTE: PBS_WALLTIME is set in seconds, but we only have HH:MM:SS-like strings
at the time of building the PBS file. We needed to add a
walltime_to_seconds helper function to convert HH:MM:SS like strings
into seconds, so that SBATCH_TIMELIMIT is set with seconds like
PBS_WALLTIME.

bouthilx committed:
Add sbatch to command-line launcher options

bouthilx committed:
Make get_launcher more flexible

It is possible to query the system to see if some commands are available
using distutils.spawn.find_executable(command_name).

Clusters where more than one launcher are available will still get
launchers selected based on string matching. For instance,
get_launcher("helios") would always return msub no matter what is
available on the system.

bouthilx committed:
Add verbosity to smart-dispatch

It is difficult to debug resuming while important process are taking place in
the pbs script automatically built by SmartDispatch.

We add verbose to smart-dispatch script and add debugging prints in epilog.

aalitaiga committed:
Updated documentation for slurm clusters

bouthilx commented on README.md:
```diff
@@ -74,3 +74,11 @@ python my_script.py 9
+Smartdispatch can also run on SLURM clusters
+All features like `--gpusPerNode` or `--coresPerNode` are supported
+However you need to pass SLURM specific features like `--qos` or `--output`as sbatch flach with:
+`smart-dispatch -q qtest@mp2 launch python my_script.py [1:4] --sbatchFlags="--qos=high -ofile.out`
```
-o Is defined by SmartDispatch, it's not a good idea to redefine it here. At least, we should not give it as a basic example.

bouthilx commented on README.md:
```diff
@@ -74,3 +74,11 @@ python my_script.py 9
+All features like `--gpusPerNode` or `--coresPerNode` are supported
+However you need to pass SLURM specific features like `--qos` or `--output`as sbatch flach with:
+`smart-dispatch -q qtest@mp2 launch python my_script.py [1:4] --sbatchFlags="--qos=high -ofile.out`
+If you use the full name of a feature you have to separate the feature and its value by a = like `--qos=high` if not you have to append the features and its value ex: `-ofile.out` is valid, not `-o file.out`
```
I feel it is a bit verbose but not so clear. What about this:
> For simplicity, --sbatchFlags supports short and long option definitions only with the following syntax:
> -Cgpu6gb or --constraint=gpu6gb
> For comparison, this would be invalid:
> -C gpu6gb or --constraint gpu6gb.

bouthilx commented on README.md:
```diff
@@ -74,3 +74,11 @@ python my_script.py 9
+
+### SLURM clusters
+
+Smartdispatch can also run on SLURM clusters
```
dot

bouthilx commented on README.md:
```diff
@@ -74,3 +74,11 @@ python my_script.py 9
+### SLURM clusters
+
+Smartdispatch can also run on SLURM clusters
+All features like `--gpusPerNode` or `--coresPerNode` are supported
```
dot again. Look at the REAME.md file on github, you will see how it is rendered.

--gpusPerCommand and --coresPerCommand would make more sense for most users than --gpusPerNode. Also, they are quite similar. I would give different ones, like --gpusPerComand and --mem for instance. Wait, do we have support for mem? Argh! let me take a look at this. :disappointed:

bouthilx committed:
Add support for SlurmJobGenerator

JobGenerators are selected by job_generator_factory based on the
cluster's name. We use a more flexible, duck typing approach for Slurm
clusters. If cluster name is not known, or not any of the if-case
clauses in the factory, then we look at which launchers are available in
the system. If it is sbatch, then a SlurmJobGenerator is built,
a JobGenerator otherwise.

bouthilx committed:
Print stderr when both qsub and sacctmgr fails

The command `sacctmgr` fails on some computers (mila01 namely), but the
current behavior gives the impression sbatch is simply not available.

Printing the stderr makes it more obvious that sbatch should be
available, but something is broken behind sacctmgr. It only appears when
using -vv options nevertheless.

bouthilx committed:
Add automatic script for cluster verification

Adding a script to do automatic verifications to assert validity of the
current code.

The verifications are not automatic unit-tests, they need automatically
checks that the process executed successfully, but the administrator
still needs to verify manually, reading the logs, that the requested
resources were provided.

Verifications can easily be combined, building on top of each others,
from complex ones to simpler ones.

Here is a list of all the verification currently implemented for slurm
clusters:

 1. very_simple_task                              (1  CPU)
 2. verify_simple_task_with_one_gpu               (1  CPU 1 GPU)
 3. verify_simple_task_with_many_gpus             (1  CPU X GPU)
 4. verify_many_task                              (X  CPU)
 5. verify_many_task_with_many_cores              (XY CPU)
 6. verify_many_task_with_one_gpu                 (X CPU X GPU)
 7. verify_many_task_with_many_gpus               (X CPU Y GPU)
 8. verify_simple_task_with_autoresume_unneeded   (1 CPU)
 9. verify_simple_task_with_autoresume_needed     (1 CPU)
10. verify_many_task_with_autoresume_needed       (X CPU)

bouthilx committed:
Add verification script for cedar

bouthilx committed:
Add verification script for graham

bouthilx committed:
Add verification script for mila

bouthilx committed:
Make get_launcher return None when no launcher

My initial though was that get_launcher should raise an error when no
launcher is found on the system since there cannot be any job launcher.
I realized that this would break the --doNotLaunch option that users may
want to use on system with no launcher, just to create the files.

aalitaiga committed:
Updated README

bouthilx committed:
Set properly account in verify_graham

The tests were failing because the account was not specified.

bouthilx committed:
Set properly account in verify_cedar

The tests were failing because the account was not specified

bouthilx committed:
Fix walltime_to_seconds convertion

There was a missing parentheses which was causing a bad conversion of
"DD:HH:MM:SS" to seconds.

The unit-test was also missing the same parentheses. I added a unit-test
to make sure such error could not occur again.
